storage/merge.go

```diff
@@ -476,29 +474,35 @@ func (c *chainSampleIterator) Next() bool {
 				heap.Push(&c.h, iter)
 			}
 		}
-
-		return len(c.h) > 0
+		if len(c.h) > 0 {
+			c.curr = heap.Pop(&c.h).(chunkenc.Iterator)
+			return true
+		}
 	}
 
 	if len(c.h) == 0 {
-		return false
+		return c.curr != nil && c.curr.Next()
 	}
 
-	currt, _ := c.At()
-	for len(c.h) > 0 {
-		nextt, _ := c.h[0].At()
-		// All but one of the overlapping samples will be dropped.
-		if nextt != currt {
-			break
+	for {
+		if c.curr.Next() {
```

Comment:
Optimization num 1: Making sure we heapify only when needed. For good case it's touching heap only for first samples.